From: Alexandru Avadanii <Alexandru.Avadanii@enea.com>
Date: Fri, 16 Sep 2016 21:31:56 +0200
Subject: [PATCH] nova: AArch64: enable hugepage support

This patch adds AArch64 to the hardcoded lists of valid
architectures for hugepage support.

Closes-bug: ARMBAND-90

Signed-off-by: Veena Lingadahalli <vlingadahalli@mvista.com>
Signed-off-by: Alexandru Avadanii <Alexandru.Avadanii@enea.com>
---
 .../nova-AArch64-enable-hugepage-support.patch     | 47 ++++++++++++++++++++++
 .../openstack_tasks/manifests/roles/compute.pp     | 12 ++++++
 2 files changed, 59 insertions(+)
 create mode 100644 deployment/puppet/openstack/files/nova-AArch64-enable-hugepage-support.patch

diff --git a/deployment/puppet/openstack/files/nova-AArch64-enable-hugepage-support.patch b/deployment/puppet/openstack/files/nova-AArch64-enable-hugepage-support.patch
new file mode 100644
index 0000000..847cab6
--- /dev/null
+++ b/deployment/puppet/openstack/files/nova-AArch64-enable-hugepage-support.patch
@@ -0,0 +1,47 @@
+From: Veena Lingadahalli <vlingadahalli@mvista.com>
+Date: Fri Sep 16 17:52:02 CEST 2016
+Subject: [PATCH] nova: AArch64: enable hugepage support
+
+This patch adds AArch64 to the hardcoded lists of valid
+architectures for hugepage support.
+
+Signed-off-by: Veena Lingadahalli <vlingadahalli@mvista.com>
+Acked-by: Alexandru Avadanii <Alexandru.Avadanii@enea.com>
+---
+
+diff --git a/compute/arch.py a/compute/arch.py
+index 265f853..6a4b4a3 100644
+--- a/compute/arch.py
++++ a/compute/arch.py
+@@ -156,6 +156,9 @@ def canonicalize(name):
+     if newname == "amd64":
+         newname = X86_64
+
++    if newname == "aarch64":
++        newname = AARCH64
++
+     if not is_valid(newname):
+         raise exception.InvalidArchitectureName(arch=name)
+
+diff --git a/virt/libvirt/driver.py a/virt/libvirt/driver.py
+index bac17cf..0e0bcae 100644
+--- a/virt/libvirt/driver.py
++++ a/virt/libvirt/driver.py
+@@ -5206,7 +5206,7 @@ class LibvirtDriver(driver.ComputeDriver):
+                     self._bad_libvirt_numa_version_warn = True
+                 return False
+
+-        support_matrix = {(arch.I686, arch.X86_64): MIN_LIBVIRT_NUMA_VERSION,
++        support_matrix = {(arch.I686, arch.X86_64, arch.AARCH64): MIN_LIBVIRT_NUMA_VERSION,
+                           (arch.PPC64,
+                            arch.PPC64LE): MIN_LIBVIRT_NUMA_VERSION_PPC}
+         caps = self._host.get_capabilities()
+@@ -5222,7 +5222,7 @@ class LibvirtDriver(driver.ComputeDriver):
+     def _has_hugepage_support(self):
+         # This means that the host can support multiple values for the size
+         # field in LibvirtConfigGuestMemoryBackingPage
+-        supported_archs = [arch.I686, arch.X86_64]
++        supported_archs = [arch.I686, arch.X86_64, arch.AARCH64]
+         caps = self._host.get_capabilities()
+         return ((caps.host.cpu.arch in supported_archs) and
+                 self._host.has_min_version(MIN_LIBVIRT_HUGEPAGE_VERSION,
diff --git a/deployment/puppet/openstack_tasks/manifests/roles/compute.pp b/deployment/puppet/openstack_tasks/manifests/roles/compute.pp
index 2f0f2a8..dc61033 100644
--- a/deployment/puppet/openstack_tasks/manifests/roles/compute.pp
+++ b/deployment/puppet/openstack_tasks/manifests/roles/compute.pp
@@ -330,6 +330,18 @@ class openstack_tasks::roles::compute {
     unless  => "patch -p1 -R -N --dry-run < ${nova_path}/libvirt-vga-console.patch",
     cwd     => $nova_path,
     require => [Package['patch']],
+  } ->
+  # FIXME(armband): Add AArch64 to arch list for nova hugepage support
+  file { "${nova_path}/nova-AArch64-enable-hugepage-support.patch":
+    ensure => "file",
+    source => "puppet:///modules/openstack/nova-AArch64-enable-hugepage-support.patch",
+  } ->
+  exec { 'nova AArch64 enable hugepage support':
+    path    => ['/usr/bin'],
+    command => "patch -p1 < ${nova_path}/nova-AArch64-enable-hugepage-support.patch",
+    unless  => "patch -p1 -R -N --dry-run < ${nova_path}/nova-AArch64-enable-hugepage-support.patch",
+    cwd     => $nova_path,
+    require => [Package['patch']],
   }

   class { '::nova::availability_zone':
